wsbroadcastserver/connectionlimiter.go

```diff
@@ -0,0 +1,153 @@
+}
+
+type ConnectionLimiterConfigFetcher func() *ConnectionLimiterConfig
+
+type ConnectionLimiter struct {
+	sync.RWMutex
+
+	ipConnectionCounts map[string]int
+	config             ConnectionLimiterConfigFetcher
+}
+
+func NewConnectionLimiter(configFetcher ConnectionLimiterConfigFetcher) *ConnectionLimiter {
+	return &ConnectionLimiter{
+		ipConnectionCounts: make(map[string]int),
+		config:             configFetcher,
+	}
+}
+
+func (l *ConnectionLimiter) IsAllowed(ip net.IP) bool {
+	l.RLock()
+	defer l.RUnlock()
+	return l.isAllowedImpl(ip)
+}
+
+func isIpv6(ip net.IP) bool {
+	// This seems to be the canonical way to distinguish IPv4 from IPv6 in Go
+	// https://stackoverflow.com/questions/22751035/golang-distinguish-ipv4-ipv6
+	// We don't care about the case where it is an IPv4 address in IPv6
+	// representation, we'll just treat that as IPv4.
+	return ip.To4() == nil
+}
+
+func (l *ConnectionLimiter) isAllowedImpl(ip net.IP) bool {
+	if ip == nil || ip.IsPrivate() || ip.IsLoopback() {
+		log.Warn("Ignoring private, looback, or unparseable IP. Please check relay and network configuration to ensure client IP addresses are detected correctly", "ip", ip)
+		return true
+	}
+
+	config := l.config()
+
+	if res := l.ipConnectionCounts[string(ip)]; res >= config.PerIpLimit {
+		return false
+	}
+
+	if isIpv6(ip) {
+		ipv6Slash48 := ip.Mask(net.CIDRMask(48, 128))
+		if ipv6Slash48 == nil {
+			log.Warn("Error taking /48 mask of ipv6 client address", "ip", ip)
+		} else if res := l.ipConnectionCounts[string(ipv6Slash48)+"/48"]; res >= config.PerIpv6Cidr48Limit {
+			return false
+		}
+
+		ipv6Slash64 := ip.Mask(net.CIDRMask(64, 128))
+		if ipv6Slash64 == nil {
+			log.Warn("Error taking /64 mask of ipv6 client address", "ip", ip)
+		} else if res := l.ipConnectionCounts[string(ipv6Slash64)+"/64"]; res >= config.PerIpv6Cidr64Limit {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (l *ConnectionLimiter) updateUsage(ip net.IP, increment bool) {
+	if ip == nil {
+		return
+	}
+
+	updateAmount := -1
+	if increment {
+		updateAmount = 1
+	}
+
+	config := l.config()
+	updateAndCheckBounds := func(ipString string, bound int) {
+		l.ipConnectionCounts[ipString] += updateAmount
+		if l.ipConnectionCounts[ipString] < 0 {
+			log.Error("BUG: Unbalanced ConnectionLimiter.updateUsage(..., false) calls")
+			l.ipConnectionCounts[ipString] = 0
+		} else if l.ipConnectionCounts[ipString] > bound {
+			log.Error("BUG: Unbalanced ConnectionLimiter.updateUsage(..., true) calls")
+			l.ipConnectionCounts[ipString] = bound
+		}
+	}
+
+	updateAndCheckBounds(string(ip), config.PerIpLimit)
+
+	if isIpv6(ip) {
+		ipv6Slash48 := ip.Mask(net.CIDRMask(48, 128))
+		if ipv6Slash48 != nil {
+			updateAndCheckBounds(string(ipv6Slash48)+"/48", config.PerIpv6Cidr48Limit)
+		}
+
+		ipv6Slash64 := ip.Mask(net.CIDRMask(64, 128))
+		if ipv6Slash64 != nil {
+			updateAndCheckBounds(string(ipv6Slash64)+"/64", config.PerIpv6Cidr64Limit)
```

Comment:
Perhaps we could pull this code out into a function that either
- returns a list of structs with the ip string and the limit
- calls a callback with the ip string and limit

So that this code could be deduplicated with `isAllowedImpl`

Reply:
I went with the list of structs idea, thanks for the suggestion.